subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskGraphExecuter.java

```diff
@@ -225,31 +227,36 @@ private void ensurePopulated() {
      */
     private class EventFiringTaskWorker implements Action<TaskInternal> {
         private final TaskExecuter taskExecuter;
-        private final BuildOperationExecutor.Operation parentOperation;
+        private final BuildOperationState parentOperation;
 
-        EventFiringTaskWorker(TaskExecuter taskExecuter, BuildOperationExecutor.Operation parentOperation) {
+        EventFiringTaskWorker(TaskExecuter taskExecuter, BuildOperationState parentOperation) {
             this.taskExecuter = taskExecuter;
             this.parentOperation = parentOperation;
         }
 
         @Override
         public void execute(final TaskInternal task) {
-            TaskOperationDescriptor taskOperation = new TaskOperationDescriptor(task);
-            BuildOperationDetails buildOperationDetails = BuildOperationDetails.displayName("Task " + task.getIdentityPath()).name(task.getIdentityPath().toString()).parent(parentOperation).operationDescriptor(taskOperation).build();
-            buildOperationExecutor.run(buildOperationDetails, new Action<BuildOperationContext>() {
+            buildOperationExecutor.setRootOperationOfCurrentThread(parentOperation);
```

Comment:
This should be one of the parameters of the run() method, like it was. A client should be able to describe the context in which work is happening directly on the work, rather than through magic static state.

Should also clear the magic static state at the completion of the work.